### PR TITLE
Formatter module py2and3

### PR DESCRIPTION
**Pull request by ashwch**

### CODE REVIEW

File: stdlib/2/formatter.pyi [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+    maxcol = ...  # type: int
+    def __init__(self, file: file = None maxcol: int = 72) -> None: ...
+    def reset(self) -> None: ...
+    def send_paragraph(self, blankline: int) -> None: ...
+    def send_line_break(self) -> None: ...
+    def send_hor_rule(self, *args, **kw) -> None: ...
+    def send_literal_data(self, data) -> None: ...
+    def send_flowing_data(self, data) -> None: ...
+
+def test(file: str = None) -> None: ...
```
Comment (JelleZijlstra) [inline stdlib/2/formatter.pyi]: This sort of thing should generally be written as `Optional[str] = ...`.

Comment (ashwch) [inline stdlib/2/formatter.pyi]: But [PEP 484](https://www.python.org/dev/peps/pep-0484/#union-types) says:

An optional type is also automatically assumed when the default value is `None` , for example:

    def handle_employee(e: Employee = None): ...

This is equivalent to:

    def handle_employee(e: Optional[Employee] = None) -> None: ...

Comment (JelleZijlstra) [inline stdlib/2/formatter.pyi]: Yes, we're likely to remove that part of PEP 484 soon. See python/typing#275.

File: stdlib/2/formatter.pyi [hunk 2]
```diff
@@ -0,0 +1,104 @@
+# Source: https://hg.python.org/cpython/file/2.7/Lib/formatter.py
+from typing import Any, IO, List, Optional, Tuple
+
+AS_IS = None
+FONT_TYPE = Tuple[str, bool, bool, bool]
```
Comment (JelleZijlstra) [inline stdlib/2/formatter.pyi]: FONT_TYPE and STYLES_TYPE don't exist at runtime; they should be defined as private. Also, preferably they shouldn't be all upper case, so `_FontType` and `_StylesType`.

File: stdlib/2/formatter.pyi [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+STYLES_TYPE = Tuple[Any, ...]
+
+class NullFormatter:
+    writer = ...  # type: Any
```
Comment (JelleZijlstra) [inline stdlib/2/formatter.pyi]: I think writer is actually of type `NullWriter`.

File: stdlib/2/formatter.pyi [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+    def __init__(self, writer: Any = None) -> None: ...
```
Comment (JelleZijlstra) [inline stdlib/2/formatter.pyi]: writer should be of type `Optional[NullWriter]` I believe. The default should be `...`.

File: stdlib/2/formatter.pyi [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+    def end_paragraph(self, blankline: int) -> None: ...
+    def add_line_break(self) -> None: ...
+    def add_hor_rule(self, *args, **kw) -> None: ...
+    def add_label_data(self, format, counter: int, blankline: Optional[int] = None) -> None: ...
```
Comment (JelleZijlstra) [inline stdlib/2/formatter.pyi]: Here and below, replace all the default values with `...`.

File: stdlib/2/formatter.pyi [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+    def pop_font(self) -> None: ...
+    def push_margin(self, margin: int) -> None: ...
+    def pop_margin(self) -> None: ...
+    def set_spacing(self, spacing: Optional[str]) -> None: ...
+    def push_style(self, *styles: STYLES_TYPE) -> None: ...
+    def pop_style(self, n: int = 1) -> None: ...
+    def assert_line_data(self, flag: int = 1) -> None: ...
+
+class AbstractFormatter:
+    writer = ...  # type: Any
```
Comment (JelleZijlstra) [inline stdlib/2/formatter.pyi]: NullWriter I think.

File: stdlib/2/formatter.pyi [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+    margin_stack = ...  # type: List[int]
+    spacing = ...  # type: Optional[str]
+    style_stack = ...  # type: Any
+    nospace = ...  # type: int
+    softspace = ...  # type: int
+    para_end = ...  # type: int
+    parskip = ...  # type: int
+    hard_break = ...  # type: int
+    have_label = ...  # type: int
+    def __init__(self, writer) -> None: ...
```
Comment (JelleZijlstra) [inline stdlib/2/formatter.pyi]: writer is of type NullWriter

File: stdlib/2/formatter.pyi [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+
+class DumbWriter(NullWriter):
+    file = ...  # type: IO
+    maxcol = ...  # type: int
+    def __init__(self, file: Optional[IO] = None, maxcol: int = 72) -> None: ...
+    def reset(self) -> None: ...
+    def send_paragraph(self, blankline: int) -> None: ...
+    def send_line_break(self) -> None: ...
+    def send_hor_rule(self, *args, **kw) -> None: ...
+    def send_literal_data(self, data) -> None: ...
```
Comment (JelleZijlstra) [inline stdlib/2/formatter.pyi]: data is a str

File: stdlib/2/formatter.pyi (hunk 3 above)
Comment (ashwch) [inline stdlib/2/formatter.pyi]: I was skeptical about using `NullWriter`. As basically we can pass any object that has methods defined in `NullWriter`, not necessarily a subclass of it.

Comment (JelleZijlstra) [inline stdlib/2/formatter.pyi]: I guess what we really need is a Protocol.

File: stdlib/2/formatter.pyi [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+    def flush_softspace(self) -> None: ...
+    def push_alignment(self, align: Optional[str]) -> None: ...
+    def pop_alignment(self) -> None: ...
+    def push_font(self, x: _FontType) -> None: ...
+    def pop_font(self) -> None: ...
+    def push_margin(self, margin: int) -> None: ...
+    def pop_margin(self) -> None: ...
+    def set_spacing(self, spacing: Optional[str]) -> None: ...
+    def push_style(self, *styles: _StylesType) -> None: ...
+    def pop_style(self, n: int = 1) -> None: ...
```
Comment (JelleZijlstra) [inline stdlib/2/formatter.pyi]: Can you change these defaults to `...` too (there's a few more 1s and a 72 below).

File: stdlib/2/formatter.pyi [hunk 10]
```diff
@@ -0,0 +1,104 @@
+# Source: https://hg.python.org/cpython/file/2.7/Lib/formatter.py
+from typing import Any, IO, List, Optional, Tuple
+
+AS_IS = None
+_FontType = Tuple[str, bool, bool, bool]
+_StylesType = Tuple[Any, ...]
+
+class NullFormatter:
+    writer = ...  # type: Optional['NullWriter']
```
Comment (JelleZijlstra) [inline stdlib/2/formatter.pyi]: You don't need to quote forward references in stubs.

**Comment by JelleZijlstra** [comment]
Thanks for all the fixes!

**Pull request merged**